integrations: Create Zoom server to server oauth integration.

laurynmm opened the pull request:
Adds an additional Zoom integration for video call options on a Zulip server using a [Zoom Server to Server OAuth app](https://developers.zoom.us/docs/internal-apps/s2s-oauth/).

Last year, Zoom changed their policies for [creating unlisted apps](https://developers.zoom.us/docs/build-flow/publishing-your-apps/#unlisted-apps) so that a self-hosted Zulip server would have to go through the app review process in order to implement the current Zoom integration, which uses a Zoom User OAuth app.

Fixes #33117.

---

**Notes**:
- Currently an [access token](https://developers.zoom.us/docs/internal-apps/s2s-oauth/#generate-access-token) is generated every time a meeting is created, but as they last for an hour (3600 seconds). We could possible store the information on redis or memcached (see [comment below](https://github.com/zulip/zulip/pull/33057#discussion_r1924500568)) and only query Zoom for a new token if the previous one expired.
- As currently written, only one of the Zoom integrations can be implemented on a Zulip server as they share the `VIDEO_ZOOM_CLIENT_ID` and `VIDEO_ZOOM_CLIENT_SECRET` settings. But we could update these changes, so that both types of Zoom app integrations could be set up on the same server by having different named settings for each integration.

---

**Screenshots and screen captures:**

<details>
<summary>Compose box error banner when user's email is not on Zoom account</summary>

![Screenshot from 2025-02-04 18-22-37](https://github.com/user-attachments/assets/c35d7ca1-4297-44bf-9490-a67b78c83692)
</details>
<details>
<summary>video call provider options</summary>

![Screenshot from 2025-01-15 19-06-27](https://github.com/user-attachments/assets/7582a4e3-6f24-4ebf-83ae-b6d70d404215)
</details>
<details>
<summary>successfully created meetings</summary>

![Screenshot from 2025-01-15 19-06-05](https://github.com/user-attachments/assets/61594ad7-e155-46a9-af01-64466b983036)
</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented on zerver/models/realms.py:
```diff
@@ -567,13 +567,20 @@ class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stub
+                "id": 5,
+            }
+        else:
+            VIDEO_CHAT_PROVIDERS["zoom"] = {
```
Based on the comment in `zproject/test_settings.py`, I believe that there would need to be some kind of workaround to be able to test this as currently written.

```python
# An important downside of the test_settings.py approach is that if we
# want to change any settings that settings.py then computes
# additional settings from (e.g. EXTERNAL_HOST), we need to do a hack
# like the below line(s) before we import from settings, for
# transmitting the value of EXTERNAL_HOST to dev_settings.py so that
# it can be set there, at the right place in the settings.py flow.
# Ick.
```

timabbott replied:
Yeah, that sounds right. Potentially the right refactor is that `VIDEO_CHAT_PROVIDERS` should have all the possible supported options, and a separated `get_enabled_video_chat_providers` function should do the computation as to which are enabled?

laurynmm replied:
Added a prep commit for that refactor.

laurynmm commented on zerver/views/video_calls.py:
```diff
@@ -185,6 +186,57 @@ def make_user_authenticated_zoom_video_call(
     return json_success(request, data={"url": res.json()["join_url"]})
 
 
+def get_zoom_server_to_server_access_token() -> str:
```
These server access tokens last an hour, so we could store them in a field on the realm and check if it's expired before generating a new one. We'd need to calculate the time we requested the token as the Zoom payload gives how long until the token expires, but not the time is was generated.

Example payload from [Zoom developer docs](https://developers.zoom.us/docs/internal-apps/s2s-oauth/#to-generate-an-access-token):
```json
{
      "access_token": "eyJhbGciOiJIUzUxMiIsInYiOiIyLjAiLCJraWQiOiI8S0lEPiJ...",
      "token_type": "bearer",
      "expires_in": 3599,
      "scope" : "user:read:admin",
      "api_url": "https://api.zoom.us"
}
```

timabbott replied:
I don't want to store an additional field on Realm; it's expensive to make that table bigger. But this would be a reasonable data structure to store in `redis` or memcached; that's a good place for transient structures valuable on that time scale.

timabbott replied:
(If we do that, we'll need to be careful to make sure we model `expires` correctly -- converting it into an expiration time and probably cutting 2 minutes because several network round-trip delays are involved and we want to just get a new one if it's almost expired).

laurynmm commented on zerver/views/video_calls.py:
```diff
@@ -185,6 +186,57 @@ def make_user_authenticated_zoom_video_call(
+            # {code: 124, message: "Access token is expired"}
+            raise InvalidZoomTokenError
+        # {code: 1001, message: "User does not exist: {email}"}
+        raise JsonableError(_("Failed to create Zoom call"))
```
For these errors, we'd want the web-app to generate a compose box banner error message, since the user would be unable to use that feature.

timabbott replied:
Yeah, that sounds right. Probably the compose box will want to just plumb through the exception returned by the server, since the error message will likely help indicate what's wrong with their configuration.

laurynmm replied:
Added a compose box banner for the case where the user's delivery email does not match a user on the configured Zoom account.

Other errors follow the pattern of the existing Zoom User OAuth integration and show the "Failed to create video call." message via `ui_report.generic_embed_error`.

laurynmm commented on zproject/default_settings.py:
```diff
@@ -119,6 +119,7 @@
 
+VIDEO_ZOOM_ACCOUNT_ID = get_secret("video_zoom_account_id", development_only=True)
 VIDEO_ZOOM_CLIENT_ID = get_secret("video_zoom_client_id", development_only=True)
 VIDEO_ZOOM_CLIENT_SECRET = get_secret("video_zoom_client_secret")
```
We could instead have different settings for this integration, e.g., `VIDEO_ZOOM_SERVER_TO_SERVER_CLIENT_ID`.

timabbott replied:
I think using the same settings for most of it is probably reasonable. But maybe `VIDEO_ZOOM_SERVER_TO_SERVER_ACCOUNT_ID` would be a good name for that detail, to hint that it's only for that specific variant. Well, it's quite verbose, but I'm not sure we have something shorter than "server_to_server" to clearly identify the new integration.

laurynmm replied:
Updated for the more verbose setting name above.

timabbott commented on zerver/tests/test_create_video_call.py:
```diff
@@ -36,187 +36,193 @@ def setUp(self) -> None:
+                responses.POST,
+                "https://zoom.us/oauth/token",
+                json={"access_token": "oldtoken", "expires_in": -60},
+            )
```
For tests, one may want to split `TestVideoCall` into having a separate `TestZoomIntegration` class, and have a subclass that just changing the `VIDEO_ZOOM_ACCOUNT_ID` in order to run the same tests with each integration. May or may not need an abstract class, depending whether one wants any tests just for one code path or the other.

laurynmm replied:
I did split out the Big Blue Button and Zoom tests to have a separate subclass for testing in a prep commit.

And then when adding the Zoom Server to Server integration, added the `override_settings` decorator from Django for the Zoom User Auth integration tests.

laurynmm commented:
**Updates**:
- Updated compose box error when the user's account email is unknown in Zoom. See updated screenshot in initial PR comment.

alya commented:
Could we try bold on the email address?

alya commented:
Does it need to say "Zoom Server to Server" in the dropdown? That doesn't seem like a user-facing term -- Googling for it produces developer docs.

Can it just say "Zoom", and then we'll document how exactly that works elsewhere?

timabbott commented:
I moved the prep commits to #33362 to merge, since those look good and have no functional impact.

timabbott commented:
> Can it just say "Zoom", and then we'll document how exactly that works elsewhere?

I don't see an obvious impediment to having the label be "Zoom" for both of the two types of Zoom integrations (The Cloud one is different), since I think we will only offer one to any given realm/organization, even though they're internally different. The main downside is that when debugging/talking with users about these, there won't be an easy way for the user to check what they're using.

alya commented:
I think it should just say "Zoom", then. We can separately ask if the server is self-hosted, and probably a server admin would need to debug in that case anyway.

laurynmm commented:
**Updates**:
- Both Zoom integrations use "Zoom" as the video call provider name, which is what's used in the dropdown menu for the video call providers. See updated screenshot below.
- Bolded the email in the error message when the user's account email does not match any email on the organization's Zoom account. See screenshot below. New string to be translated appears as:

`"Your Zulip account email (<b>{email}</b>) is not linked to this organization's Zoom account."`

- Used the `cache_with_key` decorator on `get_zoom_server_to_server_access_token` with a timeout of 56 minutes, which should expire shortly before the access token is set to expire on Zoom's end.

---

<details>
<summary>Updated compose box error message</summary>

![Screenshot from 2025-02-10 20-26-11](https://github.com/user-attachments/assets/7cfacde2-8473-441f-a219-69cf8b6ced83)
</details>
<details>
<summary>Updated dropdown options for new Zoom integration</summary>

![Screenshot from 2025-02-10 20-25-54](https://github.com/user-attachments/assets/58f021fd-e483-4002-b2bc-da1dc1f193e4)
</details>

laurynmm commented:
**Updates**:
- Added deleting the cache key for the Zoom Server to Server access token in the case that the meeting request returns an API response with the code 124, which indicates an invalid or expired access token.
- Updated the API documentation changes.
- Squashed the commits into one as the final commit only had the changes to the production doc for video call providers.

**Updated API docs**:

| Doc | Screenshot |
| --- | --- |
| API changelog entry | ![Screenshot from 2025-02-13 20-17-47](https://github.com/user-attachments/assets/270d1b3d-6e80-424c-ade2-92ec80ee9ebb) |
| Get events, https://zulip.com/api/get-events#realm-update_dict | ![Screenshot from 2025-02-13 20-18-53](https://github.com/user-attachments/assets/f38aaad8-c527-418b-99c1-fd7ad2050969) |
| Post response, https://zulip.com/api/register-queue | ![Screenshot from 2025-02-13 20-18-20](https://github.com/user-attachments/assets/f1079ce2-7a7a-4c4d-a343-b5e58640c847)

laurynmm commented on zerver/views/video_calls.py:
```diff
@@ -185,6 +198,64 @@ def make_user_authenticated_zoom_video_call(
+            # {code: 124, message: "Invalid access token"}
+            # {code: 124, message: "Access token is expired"}
+            account_id = str(settings.VIDEO_ZOOM_SERVER_TO_SERVER_ACCOUNT_ID)
+            flush_zoom_server_access_token_cache(account_id)
```
I haven't done anything with our caching system before these changes. I'm not sure if there's a better way to do this, but if we get the above documented error responses when creating a Zoom meeting, then we would want to flush any cached access token.

I imagine it might be nice for the server admin to get notified about this error, maybe as a log, so that they can check the integration?

timabbott replied:
I think logging is fine, but probably not at a level that would be an error/notify administrators.

timabbott replied:
I added the following here:

``` diff
diff --git a/zerver/tests/test_create_video_call.py b/zerver/tests/test_create_video_call.py
index 56e07d2eed..19360bae03 100644
--- a/zerver/tests/test_create_video_call.py
+++ b/zerver/tests/test_create_video_call.py
@@ -254,8 +254,18 @@ class ZoomVideoCallTestServerAuth(ZulipTestCase):
             json={"access_token": "token"},
         )
 
-        responses.add(responses.POST, self.user_zoom_meeting_url, status=400, json={"code": 124})
-        response = self.client_post("/json/calls/zoom/create")
+        responses.add(
+            responses.POST,
+            self.user_zoom_meeting_url,
+            status=400,
+            json={"code": 124, "message": "API key expired"},
+        )
+        with self.assertLogs(level="ERROR") as error_log:
+            response = self.client_post("/json/calls/zoom/create")
+            self.assertEqual(
+                error_log.output[0],
+                "ERROR:root:Unexpected Zoom error 124: API key expired",
+            )
         self.assert_json_error(response, "Failed to create Zoom call")
 
     @responses.activate
diff --git a/zerver/views/video_calls.py b/zerver/views/video_calls.py
index 001b4dfa1d..2d5cb93e02 100644
--- a/zerver/views/video_calls.py
+++ b/zerver/views/video_calls.py
@@ -1,5 +1,6 @@
 import hashlib
 import json
+import logging
 import random
 from base64 import b64encode
 from urllib.parse import quote, urlencode, urljoin
@@ -229,7 +230,8 @@ def get_zoom_server_to_server_call(
     headers = {"Authorization": f"Bearer {access_token}", "Content-Type": "application/json"}
     response = VideoCallSession().post(url, json=payload, headers=headers)
     if not response.ok:
-        zoom_api_error_code = response.json()["code"]
+        response_dict = response.json()
+        zoom_api_error_code = response_dict["code"]
         if zoom_api_error_code == 1001:
             # {code: 1001, message: "User does not exist: {email}"}
             raise UnknownZoomUserError
@@ -239,6 +241,14 @@ def get_zoom_server_to_server_call(
             # {code: 124, message: "Invalid access token"}
             # {code: 124, message: "Access token is expired"}
             account_id = str(settings.VIDEO_ZOOM_SERVER_TO_SERVER_ACCOUNT_ID)
+
+            # We are managing expiry ourselves, so this shouldn't
+            # happen. Log an error, and flush the access token from
+            # the cache, so that future requests should proceed.
+            logging.error(
+                "Unexpected Zoom error 124: %s",
+                response_dict.get("message", str(response_dict)),
+            )
             flush_zoom_server_access_token_cache(account_id)
         raise JsonableError(_("Failed to create Zoom call"))
     return response.json()["join_url"]
```

timabbott commented:
I did a bit more work on the text to try to explain why one might prefer one integration or the other. With those changes, this looks good to me, marked to merge once CI passes.

``` diff
diff --git a/docs/production/video-calls.md b/docs/production/video-calls.md
index 325b7be6dd..eb5a14ef6d 100644
--- a/docs/production/video-calls.md
+++ b/docs/production/video-calls.md
@@ -33,20 +33,29 @@ No server configuration changes are required.
 ## Zoom
 
 To use a [Zoom](https://zoom.us) integration on a self-hosted
-installation, you'll need to register a custom Zoom app.
+installation, you'll need to register a custom Zoom application for
+your Zulip server.
 
 Zulip supports two types of custom Zoom apps:
 
-- Server to Server OAuth app
-- General OAuth unlisted app
+- [Server to Server OAuth app][], which is easiest to setup, but
+  requires users to be part of the Zoom organization that created the
+  application in order to create calls.
+- [General OAuth app][], which is used by Zulip Cloud and other
+  settings where the server-to-server integration's limitations are
+  problematic.
 
 ### Server to Server OAuth app
 
-To use the Server to Server OAuth app your Zulip users' emails will need
-to match the user email addresses in your Zoom account. If a Zulip user
-is not also a user on your Zoom account, then they will not be able to
-use the integration to [create Zoom meeting links in Zulip
-messages](https://zulip.com/help/start-a-call#start-a-call).
+This Zoom application type, introduced in Zulip 10.0, is easiest to
+set up, and is ideal for most installations that self-host Zulip. To
+[create Zoom meeting links in Zulip
+messages](https://zulip.com/help/start-a-call#start-a-call) using this
+integration, users will will need to be members of your Zoom
+organization and use the same email address in Zulip that they have
+registered with Zoom.
+
+You can set up this integration as follows:
 
 1. Select [**Build App**](https://marketplace.zoom.us/develop/create)
    at the Zoom Marketplace. Create a **Server to Server OAuth App**.
@@ -67,10 +76,14 @@ messages](https://zulip.com/help/start-a-call#start-a-call).
 
 ### General OAuth app
 
-Zoom requires General OAuth apps to go through their app review process,
-even in cases where the app will not be published in their marketplace
-(see Zoom's documentation for [unlisted
-apps](https://developers.zoom.us/docs/platform/key-concepts/#private-vs-beta-vs-published-vs-unlisted-apps)).
+This Zoom application type is more flexible than the server-to-server
+integration. However, current Zoom policy requires all General OAuth
+apps to go through the full Zoom Marketplace review process, even for
+[unlisted
+apps](https://developers.zoom.us/docs/platform/key-concepts/#private-vs-beta-vs-published-vs-unlisted-apps)
+that will only be used by a single customer. As a result, you have to
+do quite a bit of publishing overhead work in order to create this
+type of Zoom application for your Zulip server.
 
 1. Select [**Build App**](https://marketplace.zoom.us/develop/create)
    at the Zoom Marketplace. Create a **General App**.
```
